core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -103,12 +103,15 @@ class MetadataCache(brokerId: Int) extends Logging {
     }
   }
 
-  def getAliveEndpoint(brokerId: Int, listenerName: ListenerName): Option[Node] =
+  private def getAliveEndpoint(brokerId: Int, listenerName: ListenerName): Option[Node] =
     inReadLock(partitionMetadataLock) {
+      // Returns None if broker is not alive or if the broker does not have a listener named `listenerName`.
+      // Since listeners can be added dynamically, a broker with a missing listener could be a transient error.
       aliveNodes.get(brokerId).map { nodeMap =>
-        nodeMap.getOrElse(listenerName,
-          throw new BrokerEndPointNotAvailableException(s"Broker `$brokerId` does not have listener with name `$listenerName`"))
```

Comment:
To check my understanding, previously when we raised this exception, the Metadata request would have failed with an unknown server error (since this exception has no error code) which would have probably been raised to the user. Is that right? Now we will return LEADER_NOT_AVAILABLE instead and the client will retry.

I am wondering in this case if we really should have a separate error code to indicate that there is no listener provided so that we can at least log a warning in the client. It seems more likely that this is the result of a misconfiguration than a delayed config update.

Reply:
Yes, we should be careful here. Otherwise we may create a hard to diagnose problem for a common case (misconfigured listener).

Reply:
@hachikuji @ijuma Thanks for the reviews.  At the moment, as Jason said, user sees an unknown server error which is not retried. Neither client nor broker has any errors in the logs to show what went wrong. I did initially consider adding a new error code for this, but the problem is that old clients wouldn't recognize the error code and I thought they wouldn't retry as a result (may be I am wrong). So I thought LEADER_NOT_AVAILABLE is a reasonable error to send to the client. Since the problem occurs only if some brokers have a listener and others don't, I was thinking it was sufficient to have a log entry in the broker logs.

Reply:
@rajinisivaram if we bumped the version of the relevant protocols in the 1.1 cycle, we could conditionally return a new error code. And fallback to `LEADER_NOT_AVAILABLE`, otherwise. If we didn't bump them, then it's less clear if it's worth doing it just for this case.

Reply:
We don't seem to have bumped up the version of `MetadataRequest` or `FindCoordinatorRequest` in 1.1.

Reply:
I have changed `warn` to `error` since it could indicate misconfiguration of a broker (will document that some errors in broker logs during reconfiguration may be transient errors since there are probably other errors too that get logged during updates).

Reply:
I looked at the metadata response handling and it seems like the current behavior is just to log a warning about any errors in the metadata response. So given that we are not failing the client anyway, I think LEADER_NOT_AVAILABLE is a better choice than UNKNOWN_SERVER_ERROR, though at least the latter has the advantage of pointing the user toward the broker logs to find the problem. It is tempting to tack on a protocol bump and a new error code to the dynamic broker config KIP to have a better solution for this problem. 

By the way, I think WARN level is actually appropriate because it does not necessarily indicate a problem and it is a bit annoying for a misconfigured client to cause a bunch of error messages on the broker.

Reply:
I guess we would raise the exception if we hit it in the FindCoordinator response though.

Reply:
@hachikuji I don't believe we can get this error with a misconfigured client, I think it should only occur with misconfigured brokers. Client has successfully made connection to a broker and that broker is looking for the listener name based on its endpoint on another broker. So perhaps `error` is ok?

Reply:
Hmm.. That's a fair point. Something else to consider is whether we should log the message when we receive the UpdateMetadata request from the broker rather than the Metadata request handler.

Reply:
@hachikuji Yes, that makes sense, updated. Do we still want to change protocol version and add a new error code for 1.1?

Reply:
@hachikuji @rajinisivaram How about we add it to the KIP but implement it in the next version? I believe we have KIPs in progress that suggest changing MetadataRequest and we could piggyback on one of them.

Reply:
That sounds good to me. I think this is still an improvement over existing behavior.

Reply:
@hachikuji @ijuma Sounds good. I will update the KIP and create a JIRA for the next version. I think all the other comments on this one have been addressed. Let me know if anything else needs to be done for 1.1. Thanks.